gui: add aggregate net descriptor

gadfort opened the pull request:
https://github.com/The-OpenROAD-Project/OpenROAD/issues/2781

maliberty commented:
This looks like a bug as the aggregate is VDD:
![image](https://user-images.githubusercontent.com/761514/215584615-e5035868-b2db-486d-a143-8c28752063c0.png)

maliberty commented on src/gui/src/aggregateNetDescriptor.cpp:
```diff
@@ -0,0 +1,223 @@
+      continue;
+    }
+
+    for (auto* next_iterm : inst->getITerms()) {
+      populate(next_iterm->getNet());
+    }
```
I expect the bug is here in not checking whether the iterm is power/gnd

gadfort replied:
Yeah, you are correct, supplies shouldn't be in the list.

maliberty replied:
if (!next_iterm->getSigType().isSupply()) {
        populate(next_iterm->getNet());
      }

seems to work

maliberty commented:
After the discussion in the issue I did come to like "Buffer tree" better than "Aggregate Net".   Perhaps it could include the buffer instances as well as the nets.

maliberty commented:
Also the non-buffer iterm/bterms would be nice (like a regular net) to see where the tree terminates.

gadfort commented:
@maliberty I renamed it to Buffer Tree, added iterms/bterms and instances to the descriptor (they are not part of the highlight though). Route guides and focus nets is also there.

maliberty commented on src/gui/src/dbDescriptors.cpp:
```diff
@@ -1332,6 +1333,10 @@ Descriptor::Properties DbNetDescriptor::getProperties(std::any object) const
     props.push_back({"Non-default rule", gui->makeSelected(ndr)});
   }
 
+  if (BufferTree::isAggregate(net)) {
+    props.push_back({"Aggregate net", gui->makeSelected(BufferTree(net))});
```
There is still some use of Aggregate in the code and visible in the GUI

maliberty commented on src/gui/src/bufferTreeDescriptor.cpp:
```diff
@@ -0,0 +1,310 @@
+{
+  populate(net);
+
+  std::sort(
+      nets_.begin(), nets_.end(), [](odb::dbNet* r, odb::dbNet* l) -> bool {
+        return r->getName() < l->getName();
+      });
+
+  name_ = nets_[0]->getName();
```
std::min_element is more efficient than sort

gadfort replied:
The sort is used to ensure a stable ordering of the nets.

maliberty commented on src/gui/src/bufferTreeDescriptor.cpp:
```diff
@@ -0,0 +1,310 @@
+    }
+  }
+
+  for (auto* iterm : net->getITerms()) {
+    auto* inst = iterm->getInst();
+    if (!BufferTree::isAggregate(inst)) {
+      if (std::find(iterm_terms_.begin(), iterm_terms_.end(), iterm)
+          == iterm_terms_.end()) {
+        iterm_terms_.push_back(iterm);
+      }
+      continue;
+    }
```
A std::set would be more efficient than linear search of a vector.

maliberty commented on src/gui/src/bufferTreeDescriptor.cpp:
```diff
@@ -0,0 +1,310 @@
+
+bool BufferTreeDescriptor::lessThan(std::any l, std::any r) const
+{
+  BufferTree* l_anet = std::any_cast<BufferTree>(&l);
+  BufferTree* r_anet = std::any_cast<BufferTree>(&r);
+  return l_anet->getName() < r_anet->getName();
```
anet needs updating

maliberty reviewed (changes requested):
I am really finding this helpful.  Thanks!

maliberty commented:
Free bonus, If you click a clock net you get the clock tree:
![image](https://user-images.githubusercontent.com/761514/215627200-7ff41421-ca8c-4c8a-9053-dc9d10d61f25.png)

gadfort commented:
Feedback should be addressed
